File: go/isogram/isogram.go
package isogram

import (
	"regexp"
	"slices"
	"strings"
)

// matches all non-alphabetic chars
var re = regexp.MustCompile(`(?mi)\W`)

// IsIsogram accepts the param "word" of type string and returns
// a boolean based on if it contains duplicate letters
func IsIsogram(word string) bool {
	// parse out non-alphabetic chars
	parsedW := re.ReplaceAllString(word, "")
	//convert the string to lowercase and split it into letters
	split := strings.Split(strings.ToLower(parsedW), "")
	// sort the letters in place
	slices.Sort(split)

	// compacts (removes consecutive isntances of an element)
	// the split and sorted string, checks its length against
	// our initial parsed string
	return len(slices.Compact(split)) == len(parsedW)
}

File: zig/isogram/isogram.zig
const std = @import("std");

/// Accepts a str and returns a bool. True if the word is an isogram, false if not
pub fn isIsogram(str: []const u8) bool {
    var chars = std.bit_set.IntegerBitSet(26).initEmpty();

    for (str) |char| {
        if (!std.ascii.isAlphabetic(char)) continue;

        const char_idx = std.ascii.toLower(char) - 'a';
        if (chars.isSet(char_idx)) return false;

        chars.set(char_idx);
    }

    return true;
}
